ceno_zkvm/src/instructions/riscv/memory/load.rs

```diff
@@ -84,6 +84,7 @@ impl<E: ExtensionField, I: RIVInstruction> Instruction<E> for LoadInstruction<E,
     ) -> Result<Self::InstructionConfig, ZKVMError> {
         let rs1_read = UInt::new_unchecked(|| "rs1_read", circuit_builder)?; // unsigned 32-bit value
         let imm = circuit_builder.create_witin(|| "imm"); // signed 12-bit value
+        // Memory initialization is not guaranteed to contain u32. Decompose in [u16; 2] here.
```

Comment:
Each memory slot MUST be initialized with u32, right?

Reply:
Actually not, the private input circuit itself will allow non-u32.

It is the LOAD or STORE that touch a private input value that will range-check it to `u32`. This was already in place, I just highlighted it with this comment.

Reply:
I still don't get it why prover want to assign non-u32 to memory slot.

Reply:
Because they’re evil and trying to make fake proofs.

Reply:
ok, i see. Load instruction needs to range check the value it reads from write set. Since both store instructions and memory init circuits are allowed to insert write record into write set, it's possible a malicious prover will try to insert some random number here.